### PR TITLE
Move REST endpoints to `POST /<RECORD>/query`

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

Structural queries currently can be passed to the `GET` endpoint. As `GET` does not support bodies, this is changed to `POST`

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1200211978612931/1202923536131153/f) _(internal)_

## 🔍 What does this change?

- Add `POST /query` endpoints
- Redirect other GET endpoints to use that endpoints
- Generated OpenAPI specs

### CODE REVIEW

File: packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs:134]: I don't think it's really important, and we shouldn't block the PR on it (in my opinion), but:

Should we think about renaming some of these to `read_*` or `retrieve_*` or something? It's a little strange to call it a `get` if it's a `post`

File: packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs:131]: Do we really not have any other error types? What if the query is malformed or invalid?

File: packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs:176]: It is quite strange to me to forward to another route implementation rather than calling the underlying thing. Especially since you're rewrapping the Expression in Json

File: packages/graph/hash_graph/lib/graph/src/api/rest/entity.rs
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/api/rest/entity.rs:151]: How does this 422 happen now?

File: packages/graph/hash_graph/lib/graph/src/api/rest/entity.rs
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/api/rest/entity.rs:154]: Just to flag this text is inconsistent with the other places (We call it Store) although I think all of the descriptions can be improved

File: packages/graph/hash_graph/lib/graph/src/api/rest/entity.rs
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/api/rest/entity.rs:153]: Inconsistent capitalisation with other descriptions

**Review by Alfred-Mountfield — CHANGES_REQUESTED** [review]
A few comments

File: packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs:131]: Currently, it only supports a return code of `500`. `404` is not supported as an empty list would then be returned (in comparison to `get_data_type`

File: packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs]: Applied in affe0f026cd75cc7464705857f2902d0a3b8d632

File: packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs]: Fixed/reverted in 6ad25fd80a13d7d2c39be7c0e75fba0b69963b08

File: packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs:131]: Would it be easier to give some better codes here? I appreciate we're revisiting error handling ASAP but if we have some available it'd be good to expose them.

The queries have some resolution failures right? If expressions aren't booleans or something. I'm assuming we'd want to treat those as 422's?

File: packages/graph/hash_graph/lib/graph/src/api/rest/entity.rs
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/api/rest/entity.rs]: Fixed in 111e317a75be960216f5c705ccf9f7a186ea7c85

File: packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs]: Hm well now it's inconsistent with some of the other ones 🙃 

I think we should either revert the commit and discuss as a team if we want to do this, or we should update all of the other `get_` methods to be called something else. (I'm not sure I'm a huge fan of 'query')

File: packages/graph/hash_graph/lib/graph/src/api/rest/entity.rs
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/api/rest/entity.rs]: I fixed the error code. This is an example output when passing an invalid entity id:
```
Invalid URL: UUID parsing failed: invalid character: expected an optional prefix of `urn:uuid:` followed by [0-9a-zA-Z], found `s` at 3
```
7bd4a83edc01ac71439853913470da9fe7fa7b10

Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/api/rest/entity.rs]: Does that correctly return a 400? I'm surprised because I don't see where we set the error code but perhaps that's just because I'm missing intellisense.

Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/api/rest/entity.rs]: Yes, it's returning 400 because `axum` will try to parse `EntityId`

File: packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs:132]: I was able to add `422` codes: 29c271ab582f0854fa8c4d5a1551a69ab906a51a

File: packages/graph/hash_graph/lib/graph/src/api/rest/entity.rs
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/api/rest/entity.rs]: Applied a hotfix: 7d4b3ae1a8420c4398f817e293887dbf64067e31

File: packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs:135]: I reverted the commit so we can take care of this later to unblock this PR.

File: packages/graph/hash_graph/tests/rest-test.http
Comment (TimDiekmann) [inline packages/graph/hash_graph/tests/rest-test.http:322]: This was a test that was not supposed to be pushed. It gots committed while rebasing.
```suggestion
GET http://127.0.0.1:4000/entities/{{person_a_entity_id}}
```

**Review by Alfred-Mountfield — APPROVED** [review]

**Pull request merged**
